pr_agent/git_providers/gitlab_provider.py

```diff
@@ -84,73 +86,108 @@ def publish_inline_comment(self, body: str, relevant_file: str, relevant_line_in
+                       'new_path': target_file.filename,
+                       'old_path': target_file.old_filename if target_file.old_filename else target_file.filename,
+                       'base_sha': d.base_commit_sha, 'start_sha': d.start_commit_sha, 'head_sha': d.head_commit_sha}
             if edit_type == 'deletion':
-                pos_obj['old_line'] = position
+                pos_obj['old_line'] = source_line_no - 1
             elif edit_type == 'addition':
-                pos_obj['new_line'] = position
+                pos_obj['new_line'] = target_line_no - 1
             else:
-                pos_obj['new_line'] = position
-                pos_obj['old_line'] = position
+                pos_obj['new_line'] = target_line_no - 1
+                pos_obj['old_line'] = source_line_no - 1
             self.mr.discussions.create({'body': body,
                                         'position': pos_obj})
 
     def publish_code_suggestion(self, body: str,
                                 relevant_file: str,
                                 relevant_lines_start: int,
                                 relevant_lines_end: int):
-        raise "not implemented yet for gitlab"
+        self.diff_files = self.diff_files if self.diff_files else self.get_diff_files()
+        target_file = None
+        for file in self.diff_files:
+            if file.filename == relevant_file:
+                if file.filename == relevant_file:
+                    target_file = file
+                    break
+        range = relevant_lines_end - relevant_lines_start + 1
+        body = body.replace('```suggestion', f'```suggestion:-0+{range}')
+
+        d = self.last_diff
+        #
+        # pos_obj = {'position_type': 'text',
+        #            'new_path': target_file.filename,
+        #            'old_path': target_file.old_filename if target_file.old_filename else target_file.filename,
+        #            'base_sha': d.base_commit_sha, 'start_sha': d.start_commit_sha, 'head_sha': d.head_commit_sha}
+        lines = target_file.head_file.splitlines()
+        relevant_line_in_file = lines[relevant_lines_start - 1]
+        edit_type, found, source_line_no, target_file, target_line_no = self.find_in_file(target_file, relevant_line_in_file)
+        self.send_inline_comment(body, edit_type, found, relevant_file, relevant_line_in_file, source_line_no,
+                                 target_file, target_line_no)
+        # if lines[relevant_lines_start][0] == '-':
+        #     pos_obj['old_line'] = relevant_lines_start
+        # elif lines[relevant_lines_start][0] == '+':
+        #     pos_obj['new_line'] = relevant_lines_start
+        # else:
+        #     pos_obj['new_line'] = relevant_lines_start
+        #     pos_obj['old_line'] = relevant_lines_start
+        # self.mr.discussions.create({'body': body,
+        #                             'position': pos_obj})
```

Comment:
**Suggestion:** The `publish_code_suggestion` method has a lot of commented out code. It's a good practice to remove commented out code to keep the codebase clean.
```suggestion
    def publish_code_suggestion(self, body: str,
                                    relevant_file: str,
                                    relevant_lines_start: int,
                                    relevant_lines_end: int):
        self.diff_files = self.diff_files if self.diff_files else self.get_diff_files()
        target_file = None
        for file in self.diff_files:
            if file.filename == relevant_file:
                if file.filename == relevant_file:
                    target_file = file
                    break
        range = relevant_lines_end - relevant_lines_start + 1
        body = body.replace('```suggestion', f'```suggestion:-0+{range}')

        d = self.last_diff
        lines = target_file.head_file.splitlines()
        relevant_line_in_file = lines[relevant_lines_start - 1]
        edit_type, found, source_line_no, target_file, target_line_no = self.find_in_file(target_file, relevant_line_in_file)
        self.send_inline_comment(body, edit_type, found, relevant_file, relevant_line_in_file, source_line_no,
                                 target_file, target_line_no)
```